core-dns pod in customer vpc failing health checks

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->


**What happened**:
We noticed an issue with customer vpc (vpc1) network: 

In a customer vpc, the core-dns pod failed to start due to failing the health check, see log below: 

As you can see from the log, the expected behavior should be: vpc1 core-dns pod should ping kubernetes-my-net-mizar-vpc1 ip address 12.2.0.2
However, vpc1 core-dns pod is pinging 10.0.0.1, which is default/Kubernetes ‘s ip address. 

**What you expected to happen**:
the expected behavior should be: vpc1 core-dns pod should ping kubernetes-my-net-mizar-vpc1 ip address 12.2.0.2

**How to reproduce it (as minimally and precisely as possible)**:
1. Create a custom vpc (called vpc1) and subnet
2. Create a arktos network that is associated with that customer vpc
3. Check the dns services
4. kubectl get pods -AT
    where it shows core-dns pods
5. kubectl --namespace kube-system logs coredns-my-net-mizar-vpc1
6. you will see health check failure but the ip address is not custom vpc's kubernetes service. Instead, it uses default kubernetes

**Anything else we need to know?**:
the default vpc (vpc0) works as designed. Only customer vpc does not work

**Environment**:
- Arktos version (0.3`):
- Cloud provider or hardware configuration: AWS
- OS (e.g: `cat /etc/os-release`): Ubuntu
- Kernel (e.g. `uname -a`): 5.6.0-rc2
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

log: 
ubuntu@ip-172-31-19-222:~/go/src/k8s.io/mizar/mizar/obj/tests$ kubectl get endpoints -AT
TENANT   NAMESPACE     NAME                         ENDPOINTS                               AGE     SERVICEGROUPID
system   default       kubernetes                   172.31.19.222:6443                      21m     1
system   kube-system   kube-dns                     20.0.0.5:53,20.0.0.5:53                 20m
system   kube-system   kube-dns-my-net-mizar-vpc0   20.0.0.2:53,20.0.0.2:53,20.0.0.2:9153   10m
system   kube-system   kube-dns-my-net-mizar-vpc1                                           7m42s
ubuntu@ip-172-31-19-222:~/go/src/k8s.io/mizar/mizar/obj/tests$ kubectl get pods -AT
TENANT   NAMESPACE     NAME                                         HASHKEY               READY   STATUS    RESTARTS   AGE
system   default       mizar-daemon-rmmlg                           7767213986685780088   1/1     Running   0          13m
system   default       mizar-operator-778b867947-t24d6              735919972758241397    1/1     Running   0          13m
system   kube-system   coredns-my-net-mizar-vpc0-7448976c4d-drfmk   980224902440174342    1/1     Running   0          11m
system   kube-system   coredns-my-net-mizar-vpc1-59756cf7d-8dk4m    5050504534201564395   0/1     Running   4          8m6s
system   kube-system   kube-dns-6fdcc959b4-qpptp                    8121308112454169817   3/3     Running   0          21m
system   kube-system   virtlet-j8pvc                                8041115816908773701   3/3     Running   0          21m
ubuntu@ip-172-31-19-222:~/go/src/k8s.io/mizar/mizar/obj/tests$ kubectl --namespace kube-system logs coredns-my-net-mizar-vpc1-59756cf7d-8dk4m
.:53
[INFO] plugin/reload: Running configuration MD5 = 201674b4a1d7573a4b37eaef9d28ac5d
CoreDNS-1.7.0
linux/amd64, go1.14.4, f59c03d
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:39818->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:44515->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:58887->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:35410->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:52075->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:34386->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:60945->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:54103->172.31.0.2:53: i/o timeout
I0918 03:24:53.401140       1 trace.go:116] Trace[911902081]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:23.4005985 +0000 UTC m=+0.020609054) (total time: 30.000511261s):
Trace[911902081]: [30.000511261s] [30.000511261s] END
E0918 03:24:53.401189       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Service: Get "https://10.0.0.1:443/api/v1/services?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
I0918 03:24:53.401141       1 trace.go:116] Trace[1427131847]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:23.400465918 +0000 UTC m=+0.020476386) (total time: 30.000581313s):
Trace[1427131847]: [30.000581313s] [30.000581313s] END
E0918 03:24:53.401213       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Namespace: Get "https://10.0.0.1:443/api/v1/namespaces?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
I0918 03:24:53.401166       1 trace.go:116] Trace[939984059]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:23.400590309 +0000 UTC m=+0.020600851) (total time: 30.000481493s):
Trace[939984059]: [30.000481493s] [30.000481493s] END
E0918 03:24:53.401233       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Endpoints: Get "https://10.0.0.1:443/api/v1/endpoints?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:53137->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:39998->172.31.0.2:53: i/o timeout
I0918 03:25:24.255846       1 trace.go:116] Trace[208240456]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:54.255148248 +0000 UTC m=+30.875158889) (total time: 30.000665277s):
Trace[208240456]: [30.000665277s] [30.000665277s] END
E0918 03:25:24.255880       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Namespace: Get "https://10.0.0.1:443/api/v1/namespaces?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
I0918 03:25:24.329615       1 trace.go:116] Trace[1106410694]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:54.329027401 +0000 UTC m=+30.949037935) (total time: 30.000556557s):
Trace[1106410694]: [30.000556557s] [30.000556557s] END
E0918 03:25:24.329641       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Endpoints: Get "https://10.0.0.1:443/api/v1/endpoints?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
I0918 03:25:24.751513       1 trace.go:116] Trace[460128162]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:54.750906411 +0000 UTC m=+31.370917032) (total time: 30.000574275s):
Trace[460128162]: [30.000574275s] [30.000574275s] END
E0918 03:25:24.751546       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Service: Get "https://10.0.0.1:443/api/v1/services?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
ubuntu@ip-172-31-19-222:~/go/src/k8s.io/mizar/mizar/obj/tests$ kubectl get services -AT
TENANT   NAMESPACE     NAME                           TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)                  AGE
system   default       kubernetes                     ClusterIP   10.0.0.1     <none>        443/TCP                  22m
system   default       kubernetes-my-net-mizar-vpc0   ClusterIP   20.0.0.9     <none>        443/TCP                  12m
system   default       kubernetes-my-net-mizar-vpc1   ClusterIP   12.2.0.2     <none>        443/TCP                  9m18s
system   kube-system   kube-dns                       ClusterIP   10.0.0.10    <none>        53/UDP,53/TCP            22m
system   kube-system   kube-dns-my-net-mizar-vpc0     ClusterIP   20.0.0.14    <none>        53/UDP,53/TCP,9153/TCP   12m
system   kube-system   kube-dns-my-net-mizar-vpc1     ClusterIP   12.2.0.5     <none>        53/UDP,53/TCP,9153/TCP   9m18s

`ubuntu@ip-172-31-19-222:~/go/src/k8s.io/mizar/mizar/obj/tests$ kubectl get endpoints -AT
TENANT   NAMESPACE     NAME                         ENDPOINTS                               AGE     SERVICEGROUPID
system   default       kubernetes                   172.31.19.222:6443                      21m     1
system   kube-system   kube-dns                     20.0.0.5:53,20.0.0.5:53                 20m
system   kube-system   kube-dns-my-net-mizar-vpc0   20.0.0.2:53,20.0.0.2:53,20.0.0.2:9153   10m
system   kube-system   kube-dns-my-net-mizar-vpc1                                           7m42s
ubuntu@ip-172-31-19-222:~/go/src/k8s.io/mizar/mizar/obj/tests$ kubectl get pods -AT
TENANT   NAMESPACE     NAME                                         HASHKEY               READY   STATUS    RESTARTS   AGE
system   default       mizar-daemon-rmmlg                           7767213986685780088   1/1     Running   0          13m
system   default       mizar-operator-778b867947-t24d6              735919972758241397    1/1     Running   0          13m
system   kube-system   coredns-my-net-mizar-vpc0-7448976c4d-drfmk   980224902440174342    1/1     Running   0          11m
system   kube-system   coredns-my-net-mizar-vpc1-59756cf7d-8dk4m    5050504534201564395   0/1     Running   4          8m6s
system   kube-system   kube-dns-6fdcc959b4-qpptp                    8121308112454169817   3/3     Running   0          21m
system   kube-system   virtlet-j8pvc                                8041115816908773701   3/3     Running   0          21m
ubuntu@ip-172-31-19-222:~/go/src/k8s.io/mizar/mizar/obj/tests$ kubectl --namespace kube-system logs coredns-my-net-mizar-vpc1-59756cf7d-8dk4m
.:53
[INFO] plugin/reload: Running configuration MD5 = 201674b4a1d7573a4b37eaef9d28ac5d
CoreDNS-1.7.0
linux/amd64, go1.14.4, f59c03d
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:39818->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:44515->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:58887->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:35410->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:52075->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:34386->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:60945->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:54103->172.31.0.2:53: i/o timeout
I0918 03:24:53.401140       1 trace.go:116] Trace[911902081]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:23.4005985 +0000 UTC m=+0.020609054) (total time: 30.000511261s):
Trace[911902081]: [30.000511261s] [30.000511261s] END
E0918 03:24:53.401189       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Service: Get "https://10.0.0.1:443/api/v1/services?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
I0918 03:24:53.401141       1 trace.go:116] Trace[1427131847]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:23.400465918 +0000 UTC m=+0.020476386) (total time: 30.000581313s):
Trace[1427131847]: [30.000581313s] [30.000581313s] END
E0918 03:24:53.401213       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Namespace: Get "https://10.0.0.1:443/api/v1/namespaces?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
I0918 03:24:53.401166       1 trace.go:116] Trace[939984059]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:23.400590309 +0000 UTC m=+0.020600851) (total time: 30.000481493s):
Trace[939984059]: [30.000481493s] [30.000481493s] END
E0918 03:24:53.401233       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Endpoints: Get "https://10.0.0.1:443/api/v1/endpoints?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:53137->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:39998->172.31.0.2:53: i/o timeout
I0918 03:25:24.255846       1 trace.go:116] Trace[208240456]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:54.255148248 +0000 UTC m=+30.875158889) (total time: 30.000665277s):
Trace[208240456]: [30.000665277s] [30.000665277s] END
E0918 03:25:24.255880       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Namespace: Get "https://10.0.0.1:443/api/v1/namespaces?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
I0918 03:25:24.329615       1 trace.go:116] Trace[1106410694]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:54.329027401 +0000 UTC m=+30.949037935) (total time: 30.000556557s):
Trace[1106410694]: [30.000556557s] [30.000556557s] END
E0918 03:25:24.329641       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Endpoints: Get "https://10.0.0.1:443/api/v1/endpoints?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
I0918 03:25:24.751513       1 trace.go:116] Trace[460128162]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:54.750906411 +0000 UTC m=+31.370917032) (total time: 30.000574275s):
Trace[460128162]: [30.000574275s] [30.000574275s] END
E0918 03:25:24.751546       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Service: Get "https://10.0.0.1:443/api/v1/services?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
ubuntu@ip-172-31-19-222:~/go/src/k8s.io/mizar/mizar/obj/tests$ kubectl get services -AT
TENANT   NAMESPACE     NAME                           TYPE        CLUSTER-IP   EXTERNAL-IP   PORT(S)                  AGE
system   default       kubernetes                     ClusterIP   10.0.0.1     <none>        443/TCP                  22m
system   default       kubernetes-my-net-mizar-vpc0   ClusterIP   20.0.0.9     <none>        443/TCP                  12m
system   default       kubernetes-my-net-mizar-vpc1   ClusterIP   12.2.0.2     <none>        443/TCP                  9m18s
system   kube-system   kube-dns                       ClusterIP   10.0.0.10    <none>        53/UDP,53/TCP            22m
system   kube-system   kube-dns-my-net-mizar-vpc0     ClusterIP   20.0.0.14    <none>        53/UDP,53/TCP,9153/TCP   12m
system   kube-system   kube-dns-my-net-mizar-vpc1     ClusterIP   12.2.0.5     <none>        53/UDP,53/TCP,9153/TCP   9m18s`

```
$ kubectl --namespace kube-system logs coredns-my-net-mizar-vpc1-59756cf7d-8dk4m
.:53
[INFO] plugin/reload: Running configuration MD5 = 201674b4a1d7573a4b37eaef9d28ac5d
CoreDNS-1.7.0
linux/amd64, go1.14.4, f59c03d
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:39818->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:44515->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:58887->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:35410->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:52075->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:34386->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:60945->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:54103->172.31.0.2:53: i/o timeout
I0918 03:24:53.401140       1 trace.go:116] Trace[911902081]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:23.4005985 +0000 UTC m=+0.020609054) (total time: 30.000511261s):
Trace[911902081]: [30.000511261s] [30.000511261s] END
E0918 03:24:53.401189       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Service: Get "https://10.0.0.1:443/api/v1/services?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
I0918 03:24:53.401141       1 trace.go:116] Trace[1427131847]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:23.400465918 +0000 UTC m=+0.020476386) (total time: 30.000581313s):
Trace[1427131847]: [30.000581313s] [30.000581313s] END
E0918 03:24:53.401213       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Namespace: Get "https://10.0.0.1:443/api/v1/namespaces?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
I0918 03:24:53.401166       1 trace.go:116] Trace[939984059]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:23.400590309 +0000 UTC m=+0.020600851) (total time: 30.000481493s):
Trace[939984059]: [30.000481493s] [30.000481493s] END
E0918 03:24:53.401233       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Endpoints: Get "https://10.0.0.1:443/api/v1/endpoints?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:53137->172.31.0.2:53: i/o timeout
[ERROR] plugin/errors: 2 2774764950619035068.7970601179162709452. HINFO: read udp 12.2.0.6:39998->172.31.0.2:53: i/o timeout
I0918 03:25:24.255846       1 trace.go:116] Trace[208240456]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:54.255148248 +0000 UTC m=+30.875158889) (total time: 30.000665277s):
Trace[208240456]: [30.000665277s] [30.000665277s] END
E0918 03:25:24.255880       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Namespace: Get "https://10.0.0.1:443/api/v1/namespaces?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
I0918 03:25:24.329615       1 trace.go:116] Trace[1106410694]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:54.329027401 +0000 UTC m=+30.949037935) (total time: 30.000556557s):
Trace[1106410694]: [30.000556557s] [30.000556557s] END
E0918 03:25:24.329641       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Endpoints: Get "https://10.0.0.1:443/api/v1/endpoints?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
I0918 03:25:24.751513       1 trace.go:116] Trace[460128162]: "Reflector ListAndWatch" name:pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125 (started: 2020-09-18 03:24:54.750906411 +0000 UTC m=+31.370917032) (total time: 30.000574275s):
Trace[460128162]: [30.000574275s] [30.000574275s] END
E0918 03:25:24.751546       1 reflector.go:178] pkg/mod/k8s.io/client-go@v0.18.3/tools/cache/reflector.go:125: Failed to list *v1.Service: Get "https://10.0.0.1:443/api/v1/services?limit=500&resourceVersion=0": dial tcp 10.0.0.1:443: i/o timeout
```

--- a/cmd/arktos-network-controller/app/controller.go
+++ b/cmd/arktos-network-controller/app/controller.go
@@ -50,30 +50,32 @@ const (
 
 // Controller represents the flat network controller
 type Controller struct {
-	domainName   string
-	cacheSynced  cache.InformerSynced
-	store        arktosv1.NetworkLister
-	queue        workqueue.RateLimitingInterface
-	netClientset *arktos.Clientset
-	svcClientset *kubernetes.Clientset
-	recorder     record.EventRecorder
+	domainName      string
+	kubeAPIServerIP string
+	cacheSynced     cache.InformerSynced
+	store           arktosv1.NetworkLister
+	queue           workqueue.RateLimitingInterface
+	netClientset    *arktos.Clientset
+	svcClientset    *kubernetes.Clientset
+	recorder        record.EventRecorder
 }
 
 // New creates the controller object
-func New(domainName string, netClientset *arktos.Clientset, svcClientset *kubernetes.Clientset, informer arktosinformer.NetworkInformer) *Controller {
+func New(domainName, kubeAPIServerIP string, netClientset *arktos.Clientset, svcClientset *kubernetes.Clientset, informer arktosinformer.NetworkInformer) *Controller {
 	utilruntime.Must(arktoscheme.AddToScheme(scheme.Scheme))
 	eventBroadcaster := record.NewBroadcaster()
 	eventBroadcaster.StartLogging(klog.Infof)
 	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: svcClientset.CoreV1().EventsWithMultiTenancy(metav1.NamespaceAll, metav1.TenantAll)})
 
 	return &Controller{
-		domainName:   domainName,
-		store:        informer.Lister(),
-		queue:        workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
-		cacheSynced:  informer.Informer().HasSynced,
-		netClientset: netClientset,
-		svcClientset: svcClientset,
-		recorder:     eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "flat-network-controller"}),
+		domainName:      domainName,
+		kubeAPIServerIP: kubeAPIServerIP,
+		store:           informer.Lister(),
+		queue:           workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		cacheSynced:     informer.Informer().HasSynced,
+		netClientset:    netClientset,
+		svcClientset:    svcClientset,
+		recorder:        eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "flat-network-controller"}),
 	}
 }
 
@@ -138,9 +140,9 @@ func (c *Controller) process(item interface{}) {
 	klog.V(5).Infof("processing network %s/%s", net.Tenant, net.Name)
 
 	if net.Spec.Type == flatNetworkType {
-		err = manageFlatNetwork(net, c.netClientset, c.svcClientset, true, c.domainName)
+		err = manageFlatNetwork(net, c.netClientset, c.svcClientset, true, c.domainName, c.kubeAPIServerIP)
 	} else {
-		err = manageNonFlatNetwork(net, c.netClientset, c.svcClientset, true, c.domainName)
+		err = manageNonFlatNetwork(net, c.netClientset, c.svcClientset, true, c.domainName, c.kubeAPIServerIP)
 	}
 
 	if err != nil {
@@ -154,7 +156,7 @@ func (c *Controller) process(item interface{}) {
 }
 
 // manageNonFlatNetwork is the core logic to manage a non-flat typed network object
-func manageNonFlatNetwork(net *v1.Network, netClient arktos.Interface, svcClient kubernetes.Interface, toDeployDNS bool, domainName string) error {
+func manageNonFlatNetwork(net *v1.Network, netClient arktos.Interface, svcClient kubernetes.Interface, toDeployDNS bool, domainName, kubeAPIServerIP string) error {
 	if net.DeletionTimestamp != nil {
 		return ensureTerminatingPhase(net, netClient)
 	}
@@ -168,7 +170,7 @@ func manageNonFlatNetwork(net *v1.Network, netClient arktos.Interface, svcClient
 	}
 
 	if toDeployDNS {
-		if err = deployDNSForNetwork(net, svcClient, domainName); err != nil {
+		if err = deployDNSForNetwork(net, svcClient, domainName, kubeAPIServerIP); err != nil {
 			return fmt.Errorf("failed to deploy per-network DNS: %v", err)
 		}
 	}
@@ -222,7 +224,7 @@ func ensureTerminatingPhase(net *v1.Network, netClient arktos.Interface) error {
 }
 
 // manageFlatNetwork is the core logic to manage a flat typed network object
-func manageFlatNetwork(net *v1.Network, netClient arktos.Interface, svcClient kubernetes.Interface, toDeployDNS bool, domainName string) error {
+func manageFlatNetwork(net *v1.Network, netClient arktos.Interface, svcClient kubernetes.Interface, toDeployDNS bool, domainName, kubeAPIServerIP string) error {
 	if net.DeletionTimestamp != nil {
 		return ensureTerminatingPhase(net, netClient)
 	}
@@ -238,7 +240,7 @@ func manageFlatNetwork(net *v1.Network, netClient arktos.Interface, svcClient ku
 	}
 
 	if toDeployDNS {
-		if err = deployDNSForNetwork(net, svcClient, domainName); err != nil {
+		if err = deployDNSForNetwork(net, svcClient, domainName, kubeAPIServerIP); err != nil {
 			return fmt.Errorf("failed to deploy per-network DNS: %v", err)
 		}
 	}

--- a/cmd/arktos-network-controller/app/controller.go
+++ b/cmd/arktos-network-controller/app/controller.go
@@ -18,6 +18,7 @@ package app
 
 import (
 	"fmt"
+	"strconv"
 	"strings"
 	"time"
 
@@ -50,32 +51,34 @@ const (
 
 // Controller represents the flat network controller
 type Controller struct {
-	domainName      string
-	kubeAPIServerIP string
-	cacheSynced     cache.InformerSynced
-	store           arktosv1.NetworkLister
-	queue           workqueue.RateLimitingInterface
-	netClientset    *arktos.Clientset
-	svcClientset    *kubernetes.Clientset
-	recorder        record.EventRecorder
+	domainName        string
+	kubeAPIServerIP   string
+	kubeAPIServerPort string
+	cacheSynced       cache.InformerSynced
+	store             arktosv1.NetworkLister
+	queue             workqueue.RateLimitingInterface
+	netClientset      *arktos.Clientset
+	svcClientset      *kubernetes.Clientset
+	recorder          record.EventRecorder
 }
 
 // New creates the controller object
-func New(domainName, kubeAPIServerIP string, netClientset *arktos.Clientset, svcClientset *kubernetes.Clientset, informer arktosinformer.NetworkInformer) *Controller {
+func New(domainName, kubeAPIServerIP string, kubeAPIServerPort int, netClientset *arktos.Clientset, svcClientset *kubernetes.Clientset, informer arktosinformer.NetworkInformer) *Controller {
 	utilruntime.Must(arktoscheme.AddToScheme(scheme.Scheme))
 	eventBroadcaster := record.NewBroadcaster()
 	eventBroadcaster.StartLogging(klog.Infof)
 	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: svcClientset.CoreV1().EventsWithMultiTenancy(metav1.NamespaceAll, metav1.TenantAll)})
 
 	return &Controller{
-		domainName:      domainName,
-		kubeAPIServerIP: kubeAPIServerIP,
-		store:           informer.Lister(),
-		queue:           workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
-		cacheSynced:     informer.Informer().HasSynced,
-		netClientset:    netClientset,
-		svcClientset:    svcClientset,
-		recorder:        eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "flat-network-controller"}),
+		domainName:        domainName,
+		kubeAPIServerIP:   kubeAPIServerIP,
+		kubeAPIServerPort: strconv.Itoa(kubeAPIServerPort),
+		store:             informer.Lister(),
+		queue:             workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		cacheSynced:       informer.Informer().HasSynced,
+		netClientset:      netClientset,
+		svcClientset:      svcClientset,
+		recorder:          eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: "flat-network-controller"}),
 	}
 }
 
@@ -140,9 +143,9 @@ func (c *Controller) process(item interface{}) {
 	klog.V(5).Infof("processing network %s/%s", net.Tenant, net.Name)
 
 	if net.Spec.Type == flatNetworkType {
-		err = manageFlatNetwork(net, c.netClientset, c.svcClientset, true, c.domainName, c.kubeAPIServerIP)
+		err = manageFlatNetwork(net, c.netClientset, c.svcClientset, true, c.domainName, c.kubeAPIServerIP, c.kubeAPIServerPort)
 	} else {
-		err = manageNonFlatNetwork(net, c.netClientset, c.svcClientset, true, c.domainName, c.kubeAPIServerIP)
+		err = manageNonFlatNetwork(net, c.netClientset, c.svcClientset, true, c.domainName, c.kubeAPIServerIP, c.kubeAPIServerPort)
 	}
 
 	if err != nil {
@@ -156,7 +159,7 @@ func (c *Controller) process(item interface{}) {
 }
 
 // manageNonFlatNetwork is the core logic to manage a non-flat typed network object
-func manageNonFlatNetwork(net *v1.Network, netClient arktos.Interface, svcClient kubernetes.Interface, toDeployDNS bool, domainName, kubeAPIServerIP string) error {
+func manageNonFlatNetwork(net *v1.Network, netClient arktos.Interface, svcClient kubernetes.Interface, toDeployDNS bool, domainName, kubeAPIServerIP, kubeAPIServerPort string) error {
 	if net.DeletionTimestamp != nil {
 		return ensureTerminatingPhase(net, netClient)
 	}
@@ -170,7 +173,7 @@ func manageNonFlatNetwork(net *v1.Network, netClient arktos.Interface, svcClient
 	}
 
 	if toDeployDNS {
-		if err = deployDNSForNetwork(net, svcClient, domainName, kubeAPIServerIP); err != nil {
+		if err = deployDNSForNetwork(net, svcClient, domainName, kubeAPIServerIP, kubeAPIServerPort); err != nil {
 			return fmt.Errorf("failed to deploy per-network DNS: %v", err)
 		}
 	}
@@ -224,7 +227,7 @@ func ensureTerminatingPhase(net *v1.Network, netClient arktos.Interface) error {
 }
 
 // manageFlatNetwork is the core logic to manage a flat typed network object
-func manageFlatNetwork(net *v1.Network, netClient arktos.Interface, svcClient kubernetes.Interface, toDeployDNS bool, domainName, kubeAPIServerIP string) error {
+func manageFlatNetwork(net *v1.Network, netClient arktos.Interface, svcClient kubernetes.Interface, toDeployDNS bool, domainName, kubeAPIServerIP, kubeAPIServerPort string) error {
 	if net.DeletionTimestamp != nil {
 		return ensureTerminatingPhase(net, netClient)
 	}
@@ -240,7 +243,7 @@ func manageFlatNetwork(net *v1.Network, netClient arktos.Interface, svcClient ku
 	}
 
 	if toDeployDNS {
-		if err = deployDNSForNetwork(net, svcClient, domainName, kubeAPIServerIP); err != nil {
+		if err = deployDNSForNetwork(net, svcClient, domainName, kubeAPIServerIP, kubeAPIServerPort); err != nil {
 			return fmt.Errorf("failed to deploy per-network DNS: %v", err)
 		}
 	}

--- a/cmd/arktos-network-controller/app/controller_test.go
+++ b/cmd/arktos-network-controller/app/controller_test.go
@@ -162,7 +162,7 @@ func TestManageFlatNetwork(t *testing.T) {
 				return true, svc, nil
 			})
 
-			err := manageFlatNetwork(tc.input, netClient, kubeClient, false, "cluster.local")
+			err := manageFlatNetwork(tc.input, netClient, kubeClient, false, "cluster.local", "192.168.0.3")
 
 			if !tc.expectingError && err != nil {
 				t.Errorf("got unexpected error: %v", err)
@@ -309,7 +309,7 @@ func TestManageNonFlatNetwork(t *testing.T) {
 				return true, svc, nil
 			})
 
-			err := manageNonFlatNetwork(tc.input, netClient, kubeClient, false, "cluster.local")
+			err := manageNonFlatNetwork(tc.input, netClient, kubeClient, false, "cluster.local", "192.168.0.3")
 
 			if !tc.expectingError && err != nil {
 				t.Errorf("got unexpected error: %v", err)
@@ -416,7 +416,7 @@ func TestNetworkPhaseShift(t *testing.T) {
 			})
 			kubeClient := fake.NewSimpleClientset()
 
-			err := manageNonFlatNetwork(tc.input, netClient, kubeClient, false, "cluster.local")
+			err := manageNonFlatNetwork(tc.input, netClient, kubeClient, false, "cluster.local", "192.168.0.3")
 			if err != nil {
 				t.Fatalf("unexpected error: %v", err)
 			}

--- a/cmd/arktos-network-controller/app/controller_test.go
+++ b/cmd/arktos-network-controller/app/controller_test.go
@@ -162,7 +162,7 @@ func TestManageFlatNetwork(t *testing.T) {
 				return true, svc, nil
 			})
 
-			err := manageFlatNetwork(tc.input, netClient, kubeClient, false, "cluster.local", "192.168.0.3")
+			err := manageFlatNetwork(tc.input, netClient, kubeClient, false, "cluster.local", "192.168.0.3", "6443")
 
 			if !tc.expectingError && err != nil {
 				t.Errorf("got unexpected error: %v", err)
@@ -309,7 +309,7 @@ func TestManageNonFlatNetwork(t *testing.T) {
 				return true, svc, nil
 			})
 
-			err := manageNonFlatNetwork(tc.input, netClient, kubeClient, false, "cluster.local", "192.168.0.3")
+			err := manageNonFlatNetwork(tc.input, netClient, kubeClient, false, "cluster.local", "192.168.0.3", "6443")
 
 			if !tc.expectingError && err != nil {
 				t.Errorf("got unexpected error: %v", err)
@@ -416,7 +416,7 @@ func TestNetworkPhaseShift(t *testing.T) {
 			})
 			kubeClient := fake.NewSimpleClientset()
 
-			err := manageNonFlatNetwork(tc.input, netClient, kubeClient, false, "cluster.local", "192.168.0.3")
+			err := manageNonFlatNetwork(tc.input, netClient, kubeClient, false, "cluster.local", "192.168.0.3", "6443")
 			if err != nil {
 				t.Fatalf("unexpected error: %v", err)
 			}

--- a/cmd/arktos-network-controller/app/dns_deployment.go
+++ b/cmd/arktos-network-controller/app/dns_deployment.go
@@ -31,7 +31,7 @@ import (
 	"k8s.io/client-go/kubernetes"
 )
 
-func deployDNSForNetwork(net *v1.Network, client kubernetes.Interface, domainName, kubeAPIServerIP string) error {
+func deployDNSForNetwork(net *v1.Network, client kubernetes.Interface, domainName, kubeAPIServerIP, kubeAPIServerPort string) error {
 	if err := ensureToCreateServiceAccount(net, client); err != nil {
 		return err
 	}
@@ -48,7 +48,7 @@ func deployDNSForNetwork(net *v1.Network, client kubernetes.Interface, domainNam
 		return err
 	}
 
-	if err := ensureToCreateDeployment(net, client, kubeAPIServerIP); err != nil {
+	if err := ensureToCreateDeployment(net, client, kubeAPIServerIP, kubeAPIServerPort); err != nil {
 		return err
 	}
 
@@ -174,7 +174,7 @@ func ensureToCreateConfigMap(net *v1.Network, client kubernetes.Interface, domai
 	return nil
 }
 
-func ensureToCreateDeployment(net *v1.Network, client kubernetes.Interface, kubeAPIServerIP string) error {
+func ensureToCreateDeployment(net *v1.Network, client kubernetes.Interface, kubeAPIServerIP, kubeAPIServerPort string) error {
 	name := dnsBaseName + "-" + net.Name
 	label := dnsServiceDefaultName + "-" + net.Name
 	configmap := dnsBaseName + "-" + net.Name
@@ -267,7 +267,7 @@ func ensureToCreateDeployment(net *v1.Network, client kubernetes.Interface, kube
 								},
 								Env: []corev1.EnvVar{
 									{Name: "KUBERNETES_SERVICE_HOST", Value: kubeAPIServerIP},
-									{Name: "KUBERNETES_SERVICE_PORT", Value: "6443"},
+									{Name: "KUBERNETES_SERVICE_PORT", Value: kubeAPIServerPort},
 								},
 								VolumeMounts: []corev1.VolumeMount{
 									{

--- a/cmd/arktos-network-controller/app/dns_deployment.go
+++ b/cmd/arktos-network-controller/app/dns_deployment.go
@@ -267,7 +267,7 @@ func ensureToCreateDeployment(net *v1.Network, client kubernetes.Interface, kube
 								},
 								Env: []corev1.EnvVar{
 									{Name: "KUBERNETES_SERVICE_HOST", Value: kubeAPIServerIP},
-									{Name: "KUBERNETES_SERVICE_PORT", Value:"6443"},
+									{Name: "KUBERNETES_SERVICE_PORT", Value: "6443"},
 								},
 								VolumeMounts: []corev1.VolumeMount{
 									{

--- a/cmd/arktos-network-controller/app/dns_deployment.go
+++ b/cmd/arktos-network-controller/app/dns_deployment.go
@@ -31,7 +31,7 @@ import (
 	"k8s.io/client-go/kubernetes"
 )
 
-func deployDNSForNetwork(net *v1.Network, client kubernetes.Interface, domainName string) error {
+func deployDNSForNetwork(net *v1.Network, client kubernetes.Interface, domainName, kubeAPIServerIP string) error {
 	if err := ensureToCreateServiceAccount(net, client); err != nil {
 		return err
 	}
@@ -48,7 +48,7 @@ func deployDNSForNetwork(net *v1.Network, client kubernetes.Interface, domainNam
 		return err
 	}
 
-	if err := ensureToCreateDeployment(net, client); err != nil {
+	if err := ensureToCreateDeployment(net, client, kubeAPIServerIP); err != nil {
 		return err
 	}
 
@@ -174,7 +174,7 @@ func ensureToCreateConfigMap(net *v1.Network, client kubernetes.Interface, domai
 	return nil
 }
 
-func ensureToCreateDeployment(net *v1.Network, client kubernetes.Interface) error {
+func ensureToCreateDeployment(net *v1.Network, client kubernetes.Interface, kubeAPIServerIP string) error {
 	name := dnsBaseName + "-" + net.Name
 	label := dnsServiceDefaultName + "-" + net.Name
 	configmap := dnsBaseName + "-" + net.Name
@@ -264,6 +264,10 @@ func ensureToCreateDeployment(net *v1.Network, client kubernetes.Interface) erro
 								Args: []string{
 									"-conf",
 									"/etc/coredns/Corefile",
+								},
+								Env: []corev1.EnvVar{
+									{Name: "KUBERNETES_SERVICE_HOST", Value: kubeAPIServerIP},
+									{Name: "KUBERNETES_SERVICE_PORT", Value:"6443"},
 								},
 								VolumeMounts: []corev1.VolumeMount{
 									{

--- a/cmd/arktos-network-controller/app/dns_deployment_test.go
+++ b/cmd/arktos-network-controller/app/dns_deployment_test.go
@@ -95,7 +95,7 @@ func TestDeployDNSForNetwork(t *testing.T) {
 		t.Run(tc.desc, func(t *testing.T) {
 			namePerNetwork := "coredns-" + tc.input.Name
 			kubeClient := fake.NewSimpleClientset(tc.objects...)
-			err := deployDNSForNetwork(tc.input, kubeClient, "cluster.local")
+			err := deployDNSForNetwork(tc.input, kubeClient, "cluster.local", "192.168.0.3")
 
 			if err != nil {
 				t.Fatalf("unexpected error: %v", err)

--- a/cmd/arktos-network-controller/app/dns_deployment_test.go
+++ b/cmd/arktos-network-controller/app/dns_deployment_test.go
@@ -95,7 +95,7 @@ func TestDeployDNSForNetwork(t *testing.T) {
 		t.Run(tc.desc, func(t *testing.T) {
 			namePerNetwork := "coredns-" + tc.input.Name
 			kubeClient := fake.NewSimpleClientset(tc.objects...)
-			err := deployDNSForNetwork(tc.input, kubeClient, "cluster.local", "192.168.0.3")
+			err := deployDNSForNetwork(tc.input, kubeClient, "cluster.local", "192.168.0.3", "6443")
 
 			if err != nil {
 				t.Fatalf("unexpected error: %v", err)

--- a/cmd/arktos-network-controller/network-controller.go
+++ b/cmd/arktos-network-controller/network-controller.go
@@ -20,6 +20,7 @@ package main
 
 import (
 	"flag"
+	"net"
 	"time"
 
 	arktosext "k8s.io/arktos-ext/pkg/generated/clientset/versioned"
@@ -31,14 +32,18 @@ import (
 	"k8s.io/kubernetes/cmd/arktos-network-controller/app"
 )
 
-const defaultWorkers = 4
+const (
+	defaultWorkers           = 4
+	defaultKubeAPIServerPort = 6443
+)
 
 var (
-	masterURL       string
-	kubeconfig      string
-	domainName      string
-	workers         int
-	kubeAPIServerIP string
+	masterURL         string
+	kubeconfig        string
+	domainName        string
+	workers           int
+	kubeAPIServerIP   string
+	kubeAPIServerPort int
 )
 
 func main() {
@@ -50,6 +55,10 @@ func main() {
 
 	if len(kubeAPIServerIP) == 0 {
 		klog.Fatalf("--kube-apiserver-ip arg must be specified in this version.")
+	}
+
+	if net.ParseIP(kubeAPIServerIP) == nil {
+		klog.Fatalf("--kube-apiserver-ip must be the valid ip address of kube-apiserver.")
 	}
 
 	defer klog.Flush()
@@ -74,7 +83,7 @@ func main() {
 	defer close(stopCh)
 
 	netInformer := informerFactory.Arktos().V1().Networks()
-	controller := app.New(domainName, kubeAPIServerIP, netClient, kubeClient, netInformer)
+	controller := app.New(domainName, kubeAPIServerIP, kubeAPIServerPort, netClient, kubeClient, netInformer)
 	netInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
 			controller.Add(obj)
@@ -95,5 +104,6 @@ func init() {
 	flag.StringVar(&masterURL, "master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
 	flag.IntVar(&workers, "concurrent-workers", defaultWorkers, "The number of workers that are allowed to process concurrently.")
 	flag.StringVar(&domainName, "cluster-domain", "cluster.local", "the cluster-internal domain name for Services.")
-	flag.StringVar(&kubeAPIServerIP, "kube-apiserver-ip", "", "the ip address kube-apiserver is listening with port 6443 on.")
+	flag.StringVar(&kubeAPIServerIP, "kube-apiserver-ip", "", "the ip address kube-apiserver is listening at.")
+	flag.IntVar(&kubeAPIServerPort, "kube-apiserver-port", defaultKubeAPIServerPort, "the port number kube-apiserver is listening on.")
 }

--- a/cmd/arktos-network-controller/network-controller.go
+++ b/cmd/arktos-network-controller/network-controller.go
@@ -49,7 +49,7 @@ func main() {
 	}
 
 	if len(kubeAPIServerIP) == 0 {
-		klog.Fatalf("--kube-apiserver arg must be specified in this version.")
+		klog.Fatalf("--kube-apiserver-ip arg must be specified in this version.")
 	}
 
 	defer klog.Flush()

--- a/cmd/arktos-network-controller/network-controller.go
+++ b/cmd/arktos-network-controller/network-controller.go
@@ -34,10 +34,11 @@ import (
 const defaultWorkers = 4
 
 var (
-	masterURL  string
-	kubeconfig string
-	domainName string
-	workers    int
+	masterURL       string
+	kubeconfig      string
+	domainName      string
+	workers         int
+	kubeAPIServerIP string
 )
 
 func main() {
@@ -45,6 +46,10 @@ func main() {
 	flag.Parse()
 	if workers <= 0 {
 		workers = defaultWorkers
+	}
+
+	if len(kubeAPIServerIP) == 0 {
+		klog.Fatalf("--kube-apiserver arg must be specified in this version.")
 	}
 
 	defer klog.Flush()
@@ -69,7 +74,7 @@ func main() {
 	defer close(stopCh)
 
 	netInformer := informerFactory.Arktos().V1().Networks()
-	controller := app.New(domainName, netClient, kubeClient, netInformer)
+	controller := app.New(domainName, kubeAPIServerIP, netClient, kubeClient, netInformer)
 	netInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
 			controller.Add(obj)
@@ -90,4 +95,5 @@ func init() {
 	flag.StringVar(&masterURL, "master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
 	flag.IntVar(&workers, "concurrent-workers", defaultWorkers, "The number of workers that are allowed to process concurrently.")
 	flag.StringVar(&domainName, "cluster-domain", "cluster.local", "the cluster-internal domain name for Services.")
+	flag.StringVar(&kubeAPIServerIP, "kube-apiserver-ip", "", "the ip address kube-apiserver is listening with port 6443 on.")
 }

--- a/docs/setup-guide/arktos-enforces-netowrk-feature.md
+++ b/docs/setup-guide/arktos-enforces-netowrk-feature.md
@@ -43,7 +43,7 @@ default   flat
 
 The current version has a limitation which requires cluster admin to specify the IP address of kube-apiserver that this controller connects to. We are working on the improvement that gets rid of this inconvenience. At this monent, please provide the ip address of the master node. Below is assuming 172.31.41.177:
 ```bash
-./_output/local/bin/linux/amd64/arktos-network-controller --kubeconfig ~/.kube/config --kube-apiserver-ip=172.31.41.177
+./_output/local/bin/linux/amd64/arktos-network-controller --kubeconfig=~/.kube/config --kube-apiserver-ip=172.31.41.177
 ```
 The config file has below content
 ```yaml

--- a/docs/setup-guide/arktos-enforces-netowrk-feature.md
+++ b/docs/setup-guide/arktos-enforces-netowrk-feature.md
@@ -40,8 +40,10 @@ default   flat
 ```
 
 4. Start the arktos-network-controller
+
+The current version has a limitation which requires cluster admin to specify the IP address of kube-apiserver that this controller connects to. We are working on the improvement that gets rid of this inconvenience. At this monent, please provide the ip address of the master node. Below is assuming 172.31.41.177:
 ```bash
-./_output/local/bin/linux/amd64/arktos-network-controller --kubeconfig ~/.kube/config
+./_output/local/bin/linux/amd64/arktos-network-controller --kubeconfig ~/.kube/config --kube-apiserver-ip=172.31.41.177
 ```
 The config file has below content
 ```yaml

